src/coreclr/tools/Common/Internal/Runtime/DehydratedData.cs

```diff
@@ -0,0 +1,116 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Debug = System.Diagnostics.Debug;
+
+namespace Internal.Runtime
+{
+    /// <summary>
+    /// Provides functionality to encode/decode dehydrated data instruction stream.
+    /// </summary>
+    /// <remarks>
+    /// The instructions use a variable length encoding and are split in two parts:
+    /// the instruction command kind and command data (payload).
+    /// The payload is an integer. Instruction kind and payload can fit into a single
+    /// byte, the encoding is one byte. Bigger payloads produce bigger instructions.
+    /// </remarks>
+    internal static class DehydratedDataCommand
+    {
+        public const byte Copy = 0x00;
+        public const byte ZeroFill = 0x01;
+        public const byte RelPtr32Reloc = 0x02;
+        public const byte PtrReloc = 0x03;
+
+        private const byte DehydratedDataCommandMask = 0x03;
+        private const int DehydratedDataCommandPayloadShift = 2;
+
+        private const int MaxRawShortPayload = (1 << (8 - DehydratedDataCommandPayloadShift)) - 1;
+        private const int MaxExtraPayloadBytes = 3;
+        private const int MaxShortPayload = MaxRawShortPayload - MaxExtraPayloadBytes;
+
+        public static byte EncodeShort(int command, int commandData)
+        {
+            Debug.Assert((command & DehydratedDataCommandMask) == command);
+            Debug.Assert(commandData <= MaxShortPayload);
+            return (byte)(command | (commandData << DehydratedDataCommandPayloadShift));
+        }
+
+        public static int Encode(int command, int commandData, byte[] buffer)
+        {
+            Debug.Assert((command & DehydratedDataCommandMask) == command);
+            int remainingData = commandData - MaxShortPayload;
+            if (remainingData <= 0)
+            {
+                buffer[0] = EncodeShort(command, commandData);
+                return 1;
+            }
+
+            int numExtraBytes = 0;
+            for (; remainingData != 0; remainingData >>= 8)
+                buffer[++numExtraBytes] = (byte)remainingData;
+            Debug.Assert(numExtraBytes <= MaxExtraPayloadBytes, "Decoder assumes this");
```

Comment:
I would be trivial to extend this to 4 billions, but I wanted to avoid the extra code (and extra impact on cache line when decoding this at runtime). My thinking was that:

* Hello world program has about 10,000 unique relocs. So we're using 0.06% of the full capacity. Hello world is about 2.8 MB in size. A program 1000 times as big as hello world would use a bit over a half of the available capacity. The final EXE would also likely be over 2 GB. At that point other things fall apart as well.
* The runs of literal data that we encode need to be short. It doesn't make sense to dehydrate a 16 MB literal run - we should have left it hydrated. The node kinds that we dehydrate don't have such unbounded growth that they could reach megabytes.

Reply:
Made it crash the compiler